### DISCUSSION TITLE
How are dockerized "steps.run" sections supposed to be written?

### DISCUSSION THREAD

**Discussion by a-recknagel** [community/community#26113] [Actions]
<p>As a context, I’m used to gitlab CI, where the general workflow of exectution block in a CI pipeline is</p>
<ul>
<li>define the docker image in which the code is supposed to be executed</li>
<li>define meta-info for the block (e.g. run conditions, artifact paths, …)</li>
<li>define the actual shell script that is going to be run</li>
</ul>
<p>I expected something similar for execution block in actions, where the equivalent sectios would be</p>
<ul>
<li>define docker image in jobs.&lt;job_id&gt;.container.image</li>
<li>define meta-info in jobs.&lt;job_id&gt;.* (excluding container.image and steps)</li>
<li>define the actual script that is going to be run</li>
</ul>
<p>The way I read it, the script would be executed given the –entrypoint of the used image. So usually some kind of shell, but maybe also python or any other script language that one can dockerize (by the way, this would be an amzing feature, and was the reason why I wanted to try out actions). But it doesn’t work that way, <a href="https://github.community/t/workflow-fails-when-running-steps-in-a-container/16123/8">because actions hacks the entrypoint before running the steps</a>.</p>
<p>Is there a good reason why? Since jobs.&lt;job_id&gt;.container.options are included before the the entrypoint override, I can’t even define a working entrypoint myself, so there is no good workaround. Or am I just approaching this tool with the wrong set of expectations?</p>

**Comment by chrispat** [comment, marked as answer]
<p>When specifying a container for the job the contianer is started by the runner and the entrypoint is overridden so the container will stay running.  We then will exec each step specified in the steps of the job inside that container using docker exec. </p>
<p>Becase we have reusable actions and multiple steps in our jobs rather than a single script the model has to be different from simply starting the container and running the single script through the entrypoint.</p>

**Comment by a-recknagel** [comment]
<p>I see, makes sense. Is alpine containers not work yet then simply a bug, and the workflow I described is in general ok?</p>

**Comment by chrispat** [comment]
<p>We are currently working on some changes to enable alpine containers to work seamlessly so your workflow in general should be just fine.</p>
<p>Update: 9/18/2019 Alpine containers are now support <a href="https://github.blog/changelog/2019-09-18-improvements-to-github-actions/" rel="nofollow noopener">https://github.blog/changelog/2019-09-18-improvements-to-github-actions/</a>.</p>
